### ISSUE TITLE
docker stats: high CPU utilization and memory leak

### ISSUE THREAD

**Issue by bprodoehl** [moby/moby#21837]
When running 80 idle containers on a 2-core DigitalOcean droplet, and then running docker stats, the daemon CPU utilization jumps to about 50-60%, and memory used by the docker daemon also climbs pretty quickly.  After stopping docker stats, the CPU usage immediately drops, but the memory never seems to get released.

**Output of `docker version`:**

```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 89
 Running: 81
 Paused: 0
 Stopped: 8
Images: 255
Server Version: 1.10.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 873
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.2.0-34-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.954 GiB
Name: host
ID: SHBD:RWEC:NZMV:NNU6:277T:ZM2B:TZEA:XGCL:3XFW:BH4L:GHKQ:SJBE
WARNING: No swap limit support
Labels:
 provider=digitalocean
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

DigitalOcean 2GB 2-core droplet.

**Steps to reproduce the issue:**
1. Launch a bunch of ubuntu containers, sitting at the bash shell
2. Run docker stats

**Describe the results you received:**
Docker daemon CPU usage consistently above 50%. Memory usage also grows, and is never reclaimed.

**Describe the results you expected:**
CPU usage <10% seems reasonable.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by bprodoehl** [comment]
This looks about the same with 1.11.0-rc4, as well.

**Comment by LK4D4** [comment]
All CPU is spent in reading and unmarshalling container state(pretty big struct) from disk. I think memory is garbage generated by json package too and should be eventually returned to os.

**Comment by xidianwlc** [comment]
when I use docker restart command the cpu load is too high. I use supervisor to manager about 20 Python process

**Comment by thaJeztah** [comment]
@xidianwlc if it's not related to `docker stats`, please open a separate issue (and more information), not in this discussion

**Comment by bprodoehl** [comment]
I think what I'd like to see for this are additional options to the /stats API endpoint to let me specify if I want CPU, I/O, memory, and network stats.  This would make it a bit easier to still use the stats endpoint with lots of containers running, if you're willing to sacrifice any of the normally-included data.  I'm thinking the query parameters would then look like this:

### Query Parameters:

- **stream** – 1/True/true or 0/False/false, pull stats once then disconnect. Default `true`.
- **cpu** – 1/True/true or 0/False/false, include CPU stats. Default `true`.
- **io** – 1/True/true or 0/False/false, include I/O stats. Default `true`.
- **memory** – 1/True/true or 0/False/false, include memory stats. Default `true`.
- **network** – 1/True/true or 0/False/false, include network stats. Default `true`.

**Comment by LK4D4** [comment]
@bprodoehl I think it's good idea.

**Comment by vincentwoo** [comment]
Has there been any focus on this? If your production application makes use of stat to monitor containers in realtime, the resource usage is very scary. Certain stats, like network usage, are very hard to gather without the use of Docker's stats API, so we have very little choice but to subscribe to all stats for all containers.

**Comment by cyberpulseraj** [comment]
Please update if any resolution about this issue

**Comment by bprodoehl** [comment]
The Docker stats API is still unusable for me.  I've been using this to watch network usage in production until this situation improves: https://github.com/bprodoehl/docker-stats-fs

So far it just handles network stats, pulling them from procfs and spitting them out in the same format as the Docker stats API.

**Comment by cpuguy83** [comment]
Here's a cpu profile from containerd with 100 containers running and running `docker stats` over a 2 minute period

![image](https://cloud.githubusercontent.com/assets/799078/17490652/d779b8d0-5d71-11e6-8d3e-39ad540b8421.png)

It's probably not exactly the best profile since it looks like containerd is starting another process (runc) to actually gather stats.

**Comment by LK4D4** [comment]
I did profiling on runc side. All CPU spent on opening and reading cgroups files, not sure what we can do here. Also, containerd spawns new runc binary for each element in stats stream.

**Comment by LK4D4** [comment]
Also containerd logic changed in 9dc2b3273db42c75368988a3885a3afd770069d9. And this change included only in Docker 1.12. I wonder if it became better or worse in 1.12 @vincentwoo @bprodoehl

**Comment by rileysaurus** [comment]
@LK4D4 I work with @vincentwoo and can say the same cpu issue occurred for us in 1.11 before we upgraded to 1.12. Although I haven't seen the cpu be as high as the issue reported is reporting, even while testing with 250+ idle containers.

**Comment by bprodoehl** [comment]
I am seeing the same high CPU utilization under 1.12.  Running docker stats with 80 containers up, both cores are at about 60%, with docker-containerd using most of that, and with 300 containers up, both cores are at about 70%.  With 80 containers, most rows of the docker stats output have values, and with 300 containers, most rows are blank.  Stopping the docker stats command, CPU usage falls down to 1%.

**Comment by LK4D4** [comment]
@bprodoehl Thanks for the info. It's pretty useful to know that CPU is spent in containerd, I'll try to profile it.

**Comment by bprodoehl** [comment]
No problem.  I know its a lot of data to chug through to get the stats on 80 or 300 containers, especially when it comes to all the memory stuff, so a solution in my mind would be to let users pare down what fields are looked at.  Should I break out my suggestion above for the added fields to the /stats API endpoint?

**Comment by LK4D4** [comment]
@bprodoehl yeah, maybe we should go that way.

**Comment by megastef** [comment]
What is the frequency for the queries in docker stats? We had a similar issue with the API with 1 second and changed the query interval to 5 or 10 seconds to reduce the load.

**Comment by LK4D4** [comment]
@megastef it's 1 second indeed.

**Comment by sevoma** [comment]
I'm running Docker 1.12.3 in Swarm mode across 2 nodes with ~400 services, 6 running at any time.
A datadog agent hits the stats api for metrics on each node.

The nodes each have 4 cores of Intel(R) Xeon(R) CPU E5-2630 0 @ 2.30GHz

`swarm-manager$: ps axo cmd,etime,time | grep [d]ockerd
/usr/bin/dockerd               12:00:53 11:03:14
`
That's a lot of CPU time; almost a whole core on average. The worker uses about half a core on average. It's necessary to do this monitoring in production, but using up to 25% of computational power to do so is a bit extreme. Are there any optimizations that can be done? I'd be happy to provide more information.

**Comment by cpuguy83** [comment]
@sevoma 
Thanks for reporting.
Issue with stats performance is known and being worked on.
Unfortunately issue goes all the way down into runc, how it collects stats, how containerd is gathering stats from runc, etc.

**Comment by loinx** [comment]
Hi *,

I have the same problem as following:
When I start the application that was dockerized on Centos 7. I got the issue when I try to run:
`docker stats` I saw that my java application was memory leak issue. But it didn't come from our code.
Currently, I couldn't find any problem inside the code.

I guess the memory leak comes from docker. So can you help me to find the solution for my problem?

Thanks,
Loi

**Comment by thaJeztah** [comment]
@loinx keep in mind that the GitHub issue tracker is not intended as a general support forum,
but for [reporting bugs and feature requests](https://raw.githubusercontent.com/moby/moby/master/.github/ISSUE_TEMPLATE.md). For other type of questions, consider using one of;

- the Docker Support Forums - https://forums.docker.com
- the [Docker community Slack channel](https://blog.docker.com/2016/11/introducing-docker-community-directory-docker-community-slack/) (register here: [http://dockr.ly/community](http://dockr.ly/community))
- [StackOverflow](https://stackoverflow.com)

**Comment by cpuguy83** [comment]
This should be largely resolved now with the containerd 1.0 integration.
I'm going to close this for now, but if you are still having issues with the latest docker release, please let me know. Thanks! 👼 🙇
